csharp/ql/src/semmle/code/csharp/AnnotatedType.qll

```diff
@@ -0,0 +1,246 @@
+    (
+      type = element.(Assignable).getType()
+      or
+      type = element.(Callable).getReturnType()
+      or
+      type = element.(Expr).getType()
+      or
+      type = element.(ArrayType).getElementType()
+      or
+      type = element.(DelegateType).getReturnType()
+    )
+  }
+}
+
+/** Holds if `bit` is a valid bit to set in the bitmask. */
+private predicate isBit(int bit) { bit = any(Annotations::TypeAnnotation a).getBit() }
+
+private int getBitMask(int bit) {
+  isBit(bit) and
+  result = 1.bitShiftLeft(bit)
+}
+
+private int getElementTypeFlags(@has_type_annotation element) {
+  result = sum(int b | type_annotation(element, b) | getBitMask(b))
+}
+
+private int getTypeArgumentFlags(ConstructedGeneric generic, int argument) {
+  exists(generic.getTypeArgument(argument)) and
+  result = sum(int b | type_argument_annotation(generic, argument, b) | getBitMask(b))
+}
+
+private int getTypeParameterFlags(TypeParameterConstraints constraints, Type type) {
+  specific_type_parameter_annotation(constraints, getTypeRef(type), _) and
+  result = sum(int b |
+      specific_type_parameter_annotation(constraints, getTypeRef(type), b)
+    |
+      getBitMask(b)
+    )
+}
+
+private newtype TAnnotatedType =
+  TAnnotatedTypeNullability(Type type, Annotations::TypeAnnotations annotations) {
+    Annotations::elementTypeAnnotations(_, type, annotations)
+    or
+    exists(ConstructedGeneric c, int i |
+      type = c.getTypeArgument(i) and
+      annotations = Annotations::TAnnotationFlags(getTypeArgumentFlags(c, i))
+    )
+    or
+    annotations = Annotations::TAnnotationFlags(getTypeParameterFlags(_, type))
+  }
+
+/** A type with additional information. */
+class AnnotatedType extends TAnnotatedType {
+  Type underlyingType;
+
+  Annotations::TypeAnnotations annotations;
+
+  AnnotatedType() { this = TAnnotatedTypeNullability(underlyingType, annotations) }
+
+  /** Gets a textual representation of this annotated type. */
+  string toString() {
+    result = annotations.getTypePrefix() + underlyingType.toString() + annotations.getTypeSuffix()
+  }
+
+  /** Gets a textual representation of this annotated type, with full type information. */
+  string toStringWithTypes() {
+    result = annotations.getTypePrefix() + underlyingType.toStringWithTypes() +
+        annotations.getTypeSuffix()
+  }
+
+  /** Gets the location of this annotated type, if any. */
+  Location getLocation() { result = underlyingType.getLocation() }
+
+  /**
+   * Gets the underlying type, for example `string` in `string?`
+   * Note that this might itself be a nullable value type (`System.Nullable`).
+   */
+  Type getUnderlyingType() { result = underlyingType }
+
+  /** Gets the type annotation set of this annotated type. */
+  Annotations::TypeAnnotations getAnnotations() { result = annotations }
+
+  /** Gets a type annotation of this annotated type. */
+  Annotations::TypeAnnotation getAnAnnotation() { result = getAnnotations().getAnAnnotation() }
```

Comment:
I think we need to make both `TypeAnnotations` and `TypeAnnotation` public given that they are exposed publicly as return types here.

Reply:
Point taken about inconsistent accessibility. My inclination would be to make the predicates `private` instead. I can't see a benefit of exposing them, and I find `AnnotatedType.isXXX` to be quite intuitive. (The general principle should be to reduce the size of the API.)